src/VecSim/algorithms/hnsw/graph_data_ram.h

```diff
@@ -0,0 +1,232 @@
+#pragma once
+
+#include "abs_graph_data.h"
+
+class IncomingEdges : private     vecsim_stl::vector<idType> ,
+                      public absEdges {
+public:
+    IncomingEdges(std::shared_ptr<VecSimAllocator> allocator) :
+        vecsim_stl::vector<idType>(allocator) {}
+
+    void
+    push(idType id) override {
+        vecsim_stl::vector<idType>::push_back(id);
+    }
+    
+    bool
+    removeIdIfExists(idType element_id) override {
+        auto it = std::find(begin(), end(), element_id);
+        if (it != end()) {
+            // Swap the last element with the current one (equivalent to removing the element id from
+            // the list).
+            *it = back();
+            pop_back();
+            return true;
+        }
+        return false;
+    }
+
+    virtual  void
+    removeId(idType element_id) override {
+        auto exists = removeIdIfExists(element_id);
+        if (!exists)
+            assert(0);
+    }
+
+    std::pair<size_t, const idType *> 
+    Get() override {
+        return {size(), data()};
+    }
+    virtual void
+    Set(std::pair<size_t , const idType *> inp) override {
+        resize(inp.first);
+        memcpy(data(), inp.second, inp.first * sizeof(idType));
+    } 
+
+    void save(std::ofstream &output);
+    void restore(std::ifstream &input);
+};
+
+
+class OutgoingEdges :  public absEdges {
+public:
+    OutgoingEdges() :num_links_(0) {}
+    ~OutgoingEdges() = default;
+    
+    void
+    push(idType id) override {
+        links_[num_links_++] = id;
+    }
+    
+    bool
+    removeIdIfExists(idType element_id) override {
+        for (size_t i = 0; i < num_links_; i++) {
+            if (links_[i] == element_id) {
+                // Swap the last element with the current one (equivalent to removing the element id from
+                // the list).
+                links_[i] = links_[num_links_-1];
+                num_links_--;
+                return true;
+            }
+        }
+        return false;
+    }
+
+    virtual  void
+    removeId(idType element_id) override {
+        auto exists = removeIdIfExists(element_id);
+        if (!exists)
+            assert(0);
+    }
```

Comment:
We do not use the remove methods today for outgoing edges, only for the incoming